examples/research_projects/codeparrot/scripts/minhash_deduplication.py

```diff
@@ -0,0 +1,207 @@
+        for element2 in extremes:
+            code2 = _shared_dataset[element2["base_index"]]["content"]
+            if jaccard_similarity(code1, code2) >= 0.85:
+                element2["copies"] += 1
+                break
+        else:
+            element1["copies"] = 1
+            extremes.append(element1)
+    return extremes
+
+
+def multipro_find_extremes(cluster_list, dataset):
+    global _shared_dataset
+    _shared_dataset = dataset
+    extremes_list = []
+    with mp.Pool() as pool:
+        for extremes in tqdm(
+            pool.imap_unordered(
+                _find_cluster_extremes_shared,
+                cluster_list,
+            ),
+            total=len(cluster_list),
+        ):
+            extremes_list.append(extremes)
+    return extremes_list
+
+
+def deduplicate_dataset(dataset: Type[Dataset]) -> Tuple[Type[Dataset], List[List[Dict]]]:
+    """Deduplicate the dataset using minhash and jaccard similarity.
+    This function first generate duplicate clusters, then each cluster
+    is reduced to the extremes that are similar to the other elements in the cluster.
+    Codes are called similar if their Jaccard similarity is greater than 0.85.
+
+    Parameters
+    ----------
+    dataset : Type[Dataset]
+        The dataset to deduplicate.
+
+    Returns
+    -------
+    ds_dedup : Type[Dataset]
+        The deduplicated dataset.
+    duplicate_clusters : List[List[Dict]]
+        The list of duplicate clusters.
+        Each cluster is a list of dicts with the following keys:
+        - base_index : int
+            The index of the code in the original dataset.
+        - repo_name : str
+        - path : str
+        - copies : int
+            The number of copies of the code in the cluster. (find_cluster_extremes)
+        - is_extreme : bool
+            Whether the code is an extreme in the cluster.
+        All the codes in the cluster are removed from the dataset except the extremes.
+    """
+    duplicate_clusters = make_duplicate_clusters(dataset)
+    duplicate_indices = set(x["base_index"] for cluster in duplicate_clusters for x in cluster)
+    # todo: make this multiprocessing, pay attention to memory usage
+    extreme_dict = {}
+    extremes_clusters = multipro_find_extremes(duplicate_clusters, dataset)
+    for extremes in extremes_clusters:
+        for element in extremes:
+            extreme_dict[element["base_index"]] = element
+    remove_indices = duplicate_indices - set(extreme_dict.keys())
+    ds_filter = dataset.filter(lambda x, idx: idx not in remove_indices, with_indices=True)
+
+    # update duplicate_clusters
+    for cluster in duplicate_clusters:
+        for element in cluster:
+            element["is_extreme"] = element["base_index"] in extreme_dict
+            if element["is_extreme"]:
+                element["copies"] = extreme_dict[element["base_index"]]["copies"]
+
+    print("Orginal dataset size: %d" % len(dataset))
```

Comment:
nit: I have a preference to use f-strings in these cases:
```suggestion
    print(f"Orginal dataset size: {len(dataset)}")
```